src/library/scala/math/BigInt.scala

```diff
@@ -110,14 +117,87 @@ object BigInt {
+
+  /** Computes the GCD of the two positive Long arguments, according to the binary GCD algorithm.
+   * Taken from the Spire library. */
+  private def longGcd(_x: Long, _y: Long): Long = {
+    if (_x == 1L || _y == 1L) return 1L
+
+    var x = _x
+    val xz = java.lang.Long.numberOfTrailingZeros(x)
+    x >>= xz
+
+    var y = _y
+    val yz = java.lang.Long.numberOfTrailingZeros(y)
+    y >>= yz
+
+    while (x != y) {
+      if (x > y) {
+        x -= y
+        x >>= java.lang.Long.numberOfTrailingZeros(x)
+      } else {
+        y -= x
+        y >>= java.lang.Long.numberOfTrailingZeros(y)
+      }
+    }
+
+    if (xz < yz) x << xz else x << yz
+  }
+
 }
 
-final class BigInt(val bigInteger: BigInteger)
+/** A type with efficient encoding of arbitrary integers.
+ *
+ * It wraps `java.math.BigInteger`, with optimization for small values that can be encoded in a `Long`.
+ */
+final class BigInt private (private var _bigInteger: BigInteger, private val _long: Long)
   extends ScalaNumber
     with ScalaNumericConversions
     with Serializable
     with Ordered[BigInt]
 {
+  // The class has a special encoding for integer that fit in a Long *and* are not equal to Long.MinValue.
+  //
+  // The Long value Long.MinValue is a tag specifying that the integer is encoded in the BigInteger field.
+  //
+  // There are three possible states for the class fields (_bigInteger, _long)
+  // 1. (null, l) where l != Long.MinValue, encodes the integer "l"
+  // 2. (b, l) where l != Long.MinValue; then b is a BigInteger with value l, encodes "l" == "b"
+  // 3a. (b, Long.MinValue) where b == Long.MinValue, encodes Long.MinValue
+  // 3b. (b, Long.MinValue) where b does not fit in a Long, encodes "b"
+  //
+  // There is only one possible transition 1. -> 2., when the method .bigInteger is called, then the field
+  // _bigInteger caches the result.
+  //
+  // The case 3a. is the only one where the BigInteger could actually fit in a Long, but as its value is used as a
+  // tag, we'll take the slow path instead.
+  //
+  // Additionally, we know that if this.isValidLong is true, then _long is the encoded value.
+
+  /** Public constructor present for compatibility. Use the BigInt.apply companion object method instead. */
+  def this(bi: BigInteger) = this(
+    bi, // even if it is a short BigInteger, we cache the instance
+    if (bi.bitLength <= 63)
+      bi.longValue // if _bigInteger is actually equal to Long.MinValue, no big deal, we'll take the slow path then
+    else
+      Long.MinValue
+  )
+
+  private def this(_long: Long) = this(if (_long == Long.MinValue) BigInt.longMinValueAsBigInteger else null, _long)
+
+  /** Returns whether the integer is encoded in the Long. Returns true for all values fitting in a Long except
+   *  Long.MinValue. */
+  private def longEncoding: Boolean = _long != Long.MinValue
+
+  def bigInteger: BigInteger = {
+    val read = _bigInteger
+    if (read ne null) _bigInteger else {
```

Comment:
why not use `if (read ne null) read` so that we don't have to read a field twice potentially?

Reply:
@viktorklang my understanding (which I admit may be flawed) is that, since another thread's write to `_bigInteger` is racy (that is, there is no HB relationship between this thread's read and the other thread's write), each read from `_bigInteger` in this thread is permitted to view the value of `_bigInteger` either before (`null`) or after the other thread's write, and it's not clear to me that the first read in this thread being non-`null` requires the second read to be as well.

Reply:
@NthPortal Indeed. reading more than once should be a problem in general, as the method gives no guarantee as to the stability of the identity of the _biginteger's value. But reading more than once would be uncessesary and slower than just returning the value in the local variable which should be in a register already.

Reply:
Corrected!